File: BalancedBinaryTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        
class Solution:
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def getDepth(root):
            if root==None:
                return 0
            ldepth = getDepth(root.left)
            if ldepth==-1:
                return -1
            rdepth = getDepth(root.right)
            if rdepth==-1:
                return -1
            if abs(ldepth-rdepth)>1:
                return -1
            return max(ldepth, rdepth)+1
            
        return getDepth(root)!=-1

if __name__=='__main__':
    root=TreeNode(1);
    a=TreeNode(2);
    b=TreeNode(3);
    c=TreeNode(4);
    d=TreeNode(5);
    e=TreeNode(6);
    
    root.left=a;
    root.right=b;
    a.left=c;
    a.right=d;
    c.left=e;
    
    solution = Solution()
    print(solution.isBalanced(root))

File: BalancedBinaryTree.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(root==NULL) return true;
        int ldepth=getDepth(root->left), rdepth=getDepth(root->right);
        if(!(ldepth-rdepth>1 || rdepth-ldepth>1) && isBalanced(root->left) && isBalanced(root->right)){
            return true;
        }else{
            return false;
        }
    }
    
    int getDepth(TreeNode *root){
        if(root==NULL) return 0;
        int ldepth=getDepth(root->left), rdepth=getDepth(root->right);
        return ldepth>rdepth?ldepth+1:rdepth+1;
    }
};

int main(){
    
    TreeNode *root=new TreeNode(1);
    TreeNode *a=new TreeNode(2);
    TreeNode *b=new TreeNode(3);
    TreeNode *c=new TreeNode(4);
    TreeNode *d=new TreeNode(5);
    TreeNode *e=new TreeNode(6);
    
    root->left=a;
    root->right=b;
    a->left=c;
    a->right=d;
    c->left=e;
    
    Solution *solution=new Solution();
    solution->isBalanced(root)?cout<<"True" : cout<<"False";
    
    return 0;
    
}
